[UI-side compositing] Keyboard scroll should not start if a wheel gesture event is in progress

rr-codes opened the pull request:
#### f72d588a05aa866762a1cf414496ad92a57f1237
<pre>
[UI-side compositing] Keyboard scroll should not start if a wheel gesture event is in progress
<a href="https://bugs.webkit.org/show_bug.cgi?id=253726">https://bugs.webkit.org/show_bug.cgi?id=253726</a>
rdar://106563114

Reviewed by Simon Fraser.

Wheel event scrolls should always take precedence over keyboard scrolls. Without UI Side Compositing,
this currently works by stopping any keyboard scrolls whenever any wheel event is received. For
keyboard scrolling with UI Side Compositing enabled, this is not an ideal mechanism. Instead, a
keyboard scroll should never begin in the first place if a wheel event gesture scroll is in progress.

There already exists ways to know if a wheel event gesture scroll is in progress on iOS, so this PR
just brings the same logic to macOS. Then, when a keyboard scroll is about to begin, it bails out
if there is a wheel event gesture scroll in progress.

This PR also improves and fixes the flakyness of the existing layout test which tests this
functionality, and makes it work when UI-side compositing is enabled or disabled.

* Source/WebCore/platform/KeyboardScrollingAnimator.cpp:
(WebCore::KeyboardScrollingAnimator::beginKeyboardScrollGesture):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidEndScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidEndScroll): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::scrollingTreeNodeWillStartScroll):
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidEndScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeWillStartScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeDidEndScroll): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingTreeIOS.cpp:
(WebKit::RemoteScrollingTreeIOS::scrollingTreeNodeWillStartScroll): Deleted.
(WebKit::RemoteScrollingTreeIOS::scrollingTreeNodeDidEndScroll): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingTreeIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::handleWheelEventPhase):

Canonical link: <a href="https://commits.webkit.org/261581@main">https://commits.webkit.org/261581@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f0ec39b035851c6ff06223bfb8c2750a00b8d53

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112050 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21178 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/670 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3830 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120704 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116111 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22528 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12327 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3638 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117811 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16735 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99895 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105090 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98691 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/446 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45723 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13598 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/476 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/93356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14274 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19649 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52475 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8074 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16067 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
```diff
@@ -132,6 +132,18 @@ void RemoteScrollingCoordinatorProxy::handleWheelEvent(const NativeWebWheelEvent
 }
 
+void RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll(ScrollingNodeID nodeID)
+{
```
We've lost the `webPageProxy().scrollingNodeScrollWillStartScroll(nodeID);` that `remotescrollingcoordinatorproxyios::scrollingTreeNodeWillStartScroll()` had. Doesn't that break things?

rr-codes replied:
fixed

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
```diff
@@ -58,6 +58,9 @@ class RemoteScrollingTree : public WebCore::ScrollingTree {
     bool scrollingTreeNodeRequestsScroll(WebCore::ScrollingNodeID, const WebCore::RequestedScrollData&) override;
     bool scrollingTreeNodeRequestsKeyboardScroll(WebCore::ScrollingNodeID, const WebCore::RequestedKeyboardScrollData&) override;
 
+    void scrollingTreeNodeWillStartScroll(WebCore::ScrollingNodeID);
```
These already exist on the base class:
```
#if PLATFORM(IOS_FAMILY)
    virtual void scrollingTreeNodeWillStartPanGesture(ScrollingNodeID) { }
    virtual void scrollingTreeNodeWillStartScroll(ScrollingNodeID) { }
    virtual void scrollingTreeNodeDidEndScroll(ScrollingNodeID) { }
#endif
```
so don't add new versions here. Move the base class ones outside the #ifdef and fix up the overrides accordingly.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
```diff
@@ -106,6 +106,22 @@ bool RemoteScrollingTree::scrollingTreeNodeRequestsKeyboardScroll(ScrollingNodeI
 
+void RemoteScrollingTree::scrollingTreeNodeWillStartScroll(ScrollingNodeID nodeID)
+{
+    RunLoop::main().dispatch([this, nodeID] {
```
I think here in the base class we should take the iOS code, which assumes no scrolling thread. Override in `RemoteScrollingTreeMac` to do the hop to the main thread.

rr-codes replied:
done

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
```diff
@@ -60,6 +60,23 @@
+    switch (phase) {
+    case PlatformWheelEventPhase::MayBegin:
+    case PlatformWheelEventPhase::Began:
+        scrollingTreeNodeWillStartScroll(nodeID);
```
We already hit `ScrollingTree::setUserScrollInProgressForNode()`, and this is adding a parallel mechanism. I think you should trigger the start/stop from `ScrollingTree::setUserScrollInProgressForNode()`.

rr-codes replied:
as in call `scrollingTreeNodeWillStartScroll` from `ScrollingTree::setUserScrollInProgressForNode`?

smfr commented on Source/WebCore/page/scrolling/ScrollingTree.cpp:
```diff
@@ -635,10 +635,13 @@ void ScrollingTree::setUserScrollInProgressForNode(ScrollingNodeID nodeID, bool
+    if (isScrolling) {
         m_treeState.nodesWithActiveUserScrolls.add(nodeID);
-    else
+        scrollingTreeNodeWillStartScroll(nodeID);
```
On iOS, `ScrollingTreeScrollingNodeDelegateIOS::scrollWillStart()` calls `ScrollingTree().scrollingTreeNodeWillStartScroll()` so will there be two calls to `scrollingTreeNodeWillStartScroll()` on iOS now?

rr-codes replied:
hmm ya probably. I'll test to confirm. If so, should I just move this new logic into ScrollingTreeMac?

rr-codes replied:
![willstartscroll](https://user-images.githubusercontent.com/16073505/224514372-f4edf8ed-409f-4aba-8ebc-de7f8281d9bd.png)

Actually, I don't think it can ever get called twice on iOS, since `setUserScrollInProgressForNode` is not called on iOS afaik.

I also tested on device on apple.com and only ever saw the method get called once.

rr-codes committed:
[UI-side compositing] Keyboard scroll should not start if a wheel gesture event is in progress
https://bugs.webkit.org/show_bug.cgi?id=253726
rdar://106563114

Reviewed by Simon Fraser.

Wheel event scrolls should always take precedence over keyboard scrolls. Without UI Side Compositing,
this currently works by stopping any keyboard scrolls whenever any wheel event is received. For
keyboard scrolling with UI Side Compositing enabled, this is not an ideal mechanism. Instead, a
keyboard scroll should never begin in the first place if a wheel event gesture scroll is in progress.

There already exists ways to know if a wheel event gesture scroll is in progress on iOS, so this PR
just brings the same logic to macOS. Then, when a keyboard scroll is about to begin, it bails out
if there is a wheel event gesture scroll in progress.

This PR also improves and fixes the flakyness of the existing layout test which tests this
functionality, and makes it work when UI-side compositing is enabled or disabled.

* Source/WebCore/platform/KeyboardScrollingAnimator.cpp:
(WebCore::KeyboardScrollingAnimator::beginKeyboardScrollGesture):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidEndScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeWillStartScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeDidEndScroll): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::scrollingTreeNodeWillStartScroll):
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidEndScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeWillStartScroll): Deleted.
(WebKit::RemoteScrollingCoordinatorProxyIOS::scrollingTreeNodeDidEndScroll): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingTreeIOS.cpp:
(WebKit::RemoteScrollingTreeIOS::scrollingTreeNodeWillStartScroll): Deleted.
(WebKit::RemoteScrollingTreeIOS::scrollingTreeNodeDidEndScroll): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingTreeIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::handleWheelEventPhase):

Canonical link: https://commits.webkit.org/261581@main
